includes/Core/Util/Entity_Factory.php

```diff
@@ -60,8 +60,54 @@ public static function from_context() {
 
 		// Otherwise, run frontend-specific `WP_Query` logic.
 		if ( $wp_the_query instanceof WP_Query ) {
-			return self::from_wp_query( $wp_the_query );
+			$entity = self::from_wp_query( $wp_the_query );
+
+			if ( null === $entity ) {
+				return null;
+			}
+
+			if ( ! defined( 'AMP__VERSION' ) ) {
+				return $entity;
+			}
+
+			$url_parts    = wp_parse_url( $_SERVER['REQUEST_URI'] ); // phpcs:ignore
+			$new_url_tail = '';
+
+			// check if the $url has amp query param.
+			if ( strpos( $url_parts['query'], 'amp' ) !== false ) {
+				$new_url_tail = '?amp=1';
+			}
+
+			// check if the $url has `/amp` in path.
+			if ( '/amp' === substr( $url_parts['path'], -strlen( '/amp' ) ) ) {
+				$new_url_tail = '/amp';
+			}
+
+			// check if the $url has `/amp/` in path.
+			if ( '/amp/' === substr( $url_parts['path'], -strlen( '/amp/' ) ) ) {
+				$new_url_tail = '/amp/';
+			}
+
+			if ( empty( $new_url_tail ) ) {
+				return $entity;
+			}
+
+			$new_url = $entity->get_url();
+			$new_url = $new_url . $new_url_tail;
+
+			$new_entity = new Entity(
+				$new_url,
+				array(
+					'id'    => $entity->get_id(),
+					'type'  => $entity->get_type(),
+					'title' => $entity->get_title(),
+					'mode'  => 'amp_secondary',
+				)
+			);
+
+			return $new_entity;
```

Comment:
Looks like this logic is almost the same for both methods. Let's extract it into a new private static method `maybe_convert_to_amp_entity( $url )` and use it in both methods.